railgun/keps/0003-kic-kubebuilder-re-architecture.md

```diff
@@ -125,6 +130,76 @@ As a contributor to the KIC I want to be able to quickly contribute new ideas an
+[cr]:https://github.com/kubernetes-sigs/controller-runtime
+[cg]:https://github.com/kubernetes/code-generator
+
+### Test Plan
+
+Prior to these efforts only minimal testing for the controller and the API functionality existed, with these efforts we will add a new integration test suite that covers:
+
+- significantly improve our unit testing coverage
+- establish integration tests in Golang which test all our APIs against real Kubernetes clusters
+- document testing requirements for new contributions going forward in CONTRIBUTING.md
+
+**NOTE**: Testing of our new Go API is covered for free by `client-gen`.
+
+#### TCPIngress Example
+
+The following is an example testing plan that outlines how we will test each of our APIs as part of this effort.
+
+##### Validation Tests
+
+- [ ] several valid permutations of each available field in the `TCPIngress` API spec are covered
+- [ ] several invalid permutations of each available field in the `TCPIngress` API spec are covered
+
+##### General Functionality Tests
+
+- [ ] `TCPIngress` resources can be created and route properly to the relevant `Service`, the route becomes available in under 1 minute
+- [ ] `TCPIngress` resources can be updated for a new backend `Service`, the routes change properly in under 1 minute
+- [ ] When `TCPIngress` spec changes, the behavior of persistent connections is tested (backend specific)
+- [ ] When `TCPIngress` spec changes, repeated parallel transient connections to the backend show no failures.
+- [ ] Long running persistent connections (where the backend server doesn't support reconnect mechanisms) disconnect and fail properly when pods change
+- [ ] `TCPIngress` resources can be deleted and the backend route is disconnected in under 1 minute
+
+##### Status Tests
+
+- [ ] `TCPIngress` status is checked for consistent `LoadBalancer` state between creation and multiple updates to the spec
+
+##### Specific Functionality Tests
+
+- [ ] `TCPIngress` ingress TLS configuration
+- [ ] `TCPIngress` ingress TLS configuration for multiple hosts
+
+##### Performance Tests
+
+- [ ] Multiple `TCPIngress` resources can be rapidly created, each backend resolving in under 1 minute
+- [ ] Bulk parallel connections through `TCPIngress` succeed, no failures at KIC/Kong level
+- [ ] Multiple `TCPIngress` resources can be rapidly deleted, each backend route becomes disconnected in under 1 minute
+
+### Graduation Criteria
+
+- [x] all APIs are ported to the new architecture
+- [x] all APIs are exposed via the public Go API
+- [x] all APIs are given a controller responsible for ingest of the relevant resources
+- [ ] the existing controller that configures the Kong Admin API is ported
+- [ ] all APIs receive new automated testing equivalent to the example provided in the above `Testing Plan` section
+- [ ] an alpha release cycle with feedback and bug reporting is conducted
+- [ ] an beta release cycle with feedback and bug reporting is conducted maintaining backwards compatibility
```

Comment:
ditto - I believe this should be a GH issue (the reason being same as above - competing "authoritative" sources of truth about requirements)

Reply:
We have milestones which cover the first beta release, would an issue really be right to cover this or do we just have another milestone where we collect bug reports during the beta release cycles?

Reply:
```suggestion
```

For now maybe we should just skip?